fix: `serialize()` circular references handling

zsilbi opened the pull request:
Issue: #114

- Added a `WeakMap` to store the serialized output of objects.
- When there is an available content for a previously referenced object, it just inserts that instead of the counter.
- Added more tests.

Currently fails to due to the bundle size constraints.

pi0 commented:
feel free to bump bundle size so we can compare.

pi0 commented:
Changes look good, i will have a double check later before merge, thanks for nice work ❤️ 

(one check I want to do is that we don't introduce much perf penalty, you can also check with bench script)

pi0 commented:
Thanks again for your help on this. considering bigger surface of changes and that this is needed mainly for equality diff, I have to think little more about it.

zsilbi commented:
> Thanks again for your help on this. considering bigger surface of changes and that this is needed mainly for equality diff, I have to think little more about it.

I extracted a version with the minimal required changes to this branch: https://github.com/zsilbi/ohash/tree/circular-references
Diff: https://github.com/unjs/ohash/compare/main...zsilbi:ohash:circular-references

All the other changes in this PR were made to reduce bundle size and improve code readability.

pi0 commented:
That's much cleaner. My main concern is why we would need to write keep and write circular values if there was not equality check conern.

zsilbi commented:
> That's much cleaner. My main concern is why we would need to write keep and write circular values if there was not equality check conern.

I ran into another issue:

```ts
import { serialize } from "ohash";
const obj = {
  a: {},
  b: new Set(),
};

obj.a = obj.b;
obj.b.add(1);
obj.b.add(obj.a);
```

pi0 commented:
would you mind to open another PR (or issue) for it? I agree we should pass the same context no nested serialize.

zsilbi commented:
> would you mind to open another PR (or issue) for it? I agree we should pass the same context no nested serialize.

```ts
import { serialize } from "ohash";

serialize(new Set([{ b: 1 }, { a: 1 }])) // Expected: "Set[{a:1},{b:1}]"
```

When you run this and the comparison function runs within the same context, it will store the counters first in the context and it will become: `"Set[#1,#2]"`

I am afraid that this PR might be needed to properly handle this, so I opened an issue: https://github.com/unjs/ohash/issues/122

zsilbi commented:
I removed all to cosmetic changes from this branch so the diff should be cleaner.

I had to remove the buffer because although it passed all previous tests the output was not accurate, so I added a new test for this. Therefore the performance penalty is back, because it has to `slice()` the serialized output for each object.

A better solution would be to store only the start/end positions of the object content in the context, and slice the output when the content is needed.  For plain objects there would be no overhead. 

The problem with this is that we cannot reuse the same context for the serializations within `compare()` because the context would be unique for each `Serializer`.

zsilbi commented:
I ran some benchmarks [here](https://github.com/unjs/ohash/tree/f15010df7a939aaec592011b5d19d36d7ca7fb91):

- ❌ **serialize v0 @ 7529f93**: uses a buffer, it is fast but the output is not correct 
- ❌ **serialize v1 @ 461e7f3**: does string slicing for each object, it is very slow in non-ref tests
- ❌ **serialize v2 @ 73d01b6**: stores positions, does string slicing only in child contexts, it is very slow in circular tests with large objects
- ✅ **serialize v3 @ 48f1ab4**: returns the written value of each function, so no slicing is needed, adds only a minimal overhead 

Combined results: 

```scala
 ✓ test/benchmarks.bench.ts > benchmarks > combined 8927ms
     name                        hz     min      max    mean     p75      p99     p995     p999      rme  samples
   · serialize @ main        125.83  6.9012  11.0978  7.9471  8.3344  11.0978  11.0978  11.0978   ±2.54%       63
   · serialize v0 @ 7529f93  132.32  6.7958   8.6080  7.5575  7.8211   8.6080   8.6080   8.6080   ±1.34%       67   fastest
   · serialize v1 @ 461e7f3  3.7415  194.78   342.04  267.27  303.09   342.04   342.04   342.04  ±13.26%       10   slowest
   · serialize v2 @ 73d01b6  6.5189  109.49   244.53  153.40  167.51   244.53   244.53   244.53  ±22.93%       10
   · serialize v3 @ 48f1ab4  124.25  6.9112   9.5367  8.0482  8.4230   9.5367   9.5367   9.5367   ±1.93%       63

 BENCH  Summary

  serialize v0 @ 7529f93 - test/benchmarks.bench.ts > benchmarks > combined
    1.05x faster than serialize @ main
    1.06x faster than serialize v3 @ 48f1ab4
    20.30x faster than serialize v2 @ 73d01b6
    35.36x faster than serialize v1 @ 461e7f3
```


<details>
  <summary>Detailed results</summary>
 
  ```scala
 ✓ test/benchmarks.bench.ts > benchmarks > serialize 1 object 3617ms
     name                            hz     min     max    mean     p75     p99    p995    p999     rme  samples
   · serialize @ main        597,859.79  0.0014  1.2951  0.0017  0.0016  0.0030  0.0041  0.0140  ±0.69%   298930
   · serialize v0 @ 7529f93  625,111.70  0.0014  0.2070  0.0016  0.0016  0.0025  0.0027  0.0098  ±0.18%   312556   fastest
   · serialize v1 @ 461e7f3  590,420.60  0.0015  0.2789  0.0017  0.0016  0.0029  0.0031  0.0080  ±0.44%   295211   slowest
   · serialize v2 @ 73d01b6  592,713.79  0.0014  0.3948  0.0017  0.0016  0.0026  0.0028  0.0111  ±0.53%   296357
   · serialize v3 @ 48f1ab4  595,651.16  0.0015  0.3446  0.0017  0.0017  0.0026  0.0029  0.0108  ±0.51%   297826

 ✓ test/benchmarks.bench.ts > benchmarks > serialize array with 1000 identical small objects 3152ms
     name                         hz      min      max     mean      p75      p99     p995     p999      rme  samples
   · serialize @ main         625.22   1.4046   9.0759   1.5994   1.5354   2.8562   3.0640   9.0759   ±3.48%      313   fastest
   · serialize v0 @ 7529f93   623.50   1.4645   2.5250   1.6038   1.5839   2.4159   2.5136   2.5250   ±1.42%      312
   · serialize v1 @ 461e7f3  39.2353  13.2147  40.6101  25.4873  28.3782  40.6101  40.6101  40.6101  ±12.47%       20   slowest
   · serialize v2 @ 73d01b6   620.87   1.4305   2.7438   1.6106   1.5254   2.5273   2.6303   2.7438   ±1.94%      311
   · serialize v3 @ 48f1ab4   613.13   1.4590   3.1195   1.6310   1.5735   2.6572   2.8334   3.1195   ±1.88%      307

 ✓ test/benchmarks.bench.ts > benchmarks > serialize array with 1000 identical small objects (by ref) 3023ms
     name                           hz     min     max    mean     p75     p99    p995    p999     rme  samples
   · serialize @ main        13,615.14  0.0654  0.5471  0.0734  0.0716  0.2249  0.3107  0.4070  ±0.86%     6808
   · serialize v0 @ 7529f93  15,610.49  0.0587  0.6373  0.0641  0.0611  0.1102  0.2403  0.4426  ±0.82%     7806
   · serialize v1 @ 461e7f3  16,369.38  0.0559  0.5808  0.0611  0.0582  0.0978  0.3251  0.4803  ±0.99%     8185   fastest
   · serialize v2 @ 73d01b6  14,124.18  0.0569  0.6468  0.0708  0.0697  0.1012  0.2798  0.4046  ±0.78%     7063
   · serialize v3 @ 48f1ab4  13,329.10  0.0591  0.6340  0.0750  0.0728  0.2516  0.3208  0.4032  ±0.92%     6665   slowest

 ✓ test/benchmarks.bench.ts > benchmarks > serialize array with 100 identical large circular objects 6786ms
     name                        hz      min     max    mean     p75     p99    p995    p999      rme  samples
   · serialize @ main        233.74   3.6018  6.0772  4.2782  4.8266  5.9529  6.0772  6.0772   ±2.90%      117
   · serialize v0 @ 7529f93  244.61   3.4817  5.7671  4.0882  4.7040  5.2622  5.7671  5.7671   ±2.46%      123   fastest
   · serialize v1 @ 461e7f3  6.6668   107.46  272.41  150.00  191.59  272.41  272.41  272.41  ±25.63%       10   slowest
   · serialize v2 @ 73d01b6  7.7941  99.0986  169.31  128.30  145.87  169.31  169.31  169.31  ±13.42%       10
   · serialize v3 @ 48f1ab4  199.13   3.6301  7.6277  5.0218  6.2832  7.5890  7.6277  7.6277   ±5.11%      100

 ✓ test/benchmarks.bench.ts > benchmarks > serialize array with 100 identical large objects (by ref) 3039ms
     name                           hz     min     max    mean     p75     p99    p995    p999      rme  samples
   · serialize @ main        19,453.81  0.0337  108.20  0.0514  0.0372  0.0853  0.1085  0.6094  ±42.47%     9727   slowest
   · serialize v0 @ 7529f93  25,329.58  0.0335  2.0991  0.0395  0.0368  0.0749  0.0963  0.6586   ±1.68%    12666   fastest
   · serialize v1 @ 461e7f3  23,312.16  0.0358  2.4449  0.0429  0.0403  0.0833  0.1144  0.5805   ±1.63%    11657
   · serialize v2 @ 73d01b6  24,071.65  0.0342  2.5158  0.0415  0.0381  0.0909  0.1075  0.6465   ±1.75%    12036
   · serialize v3 @ 48f1ab4  25,149.36  0.0331  2.4115  0.0398  0.0365  0.0796  0.1090  0.6765   ±1.79%    12575

 ✓ test/benchmarks.bench.ts > benchmarks > serialize array with 100 identical large objects 3503ms
     name                         hz      min     max     mean      p75     p99    p995    p999      rme  samples
   · serialize @ main         328.22   2.6025  4.6129   3.0467   3.3024  4.5599  4.6129  4.6129   ±2.38%      165   fastest
   · serialize v0 @ 7529f93   310.31   2.6460  7.0603   3.2226   3.6126  5.4400  7.0603  7.0603   ±3.21%      156
   · serialize v1 @ 461e7f3  15.6481  41.5774  102.84  63.9057  66.6241  102.84  102.84  102.84  ±19.83%       10   slowest
   · serialize v2 @ 73d01b6   314.86   2.6572  5.3061   3.1760   3.6431  5.1662  5.3061  5.3061   ±3.12%      158
   · serialize v3 @ 48f1ab4   299.32   2.6693  5.6847   3.3409   3.9853  5.3358  5.6847  5.6847   ±3.67%      150

 ✓ test/benchmarks.bench.ts > benchmarks > serialize array with 100 identical large circular objects (by ref) 3032ms
     name                           hz     min     max    mean     p75     p99    p995    p999     rme  samples
   · serialize @ main        18,436.40  0.0432  1.7740  0.0542  0.0487  0.0990  0.1420  1.3888  ±2.95%     9219
   · serialize v0 @ 7529f93  19,020.55  0.0414  1.8070  0.0526  0.0467  0.0973  0.1225  1.4893  ±3.25%     9511
   · serialize v1 @ 461e7f3  16,643.90  0.0454  2.0383  0.0601  0.0523  0.1147  1.0725  1.5446  ±3.50%     8322
   · serialize v2 @ 73d01b6  15,801.81  0.0469  8.4129  0.0633  0.0533  0.1309  1.0324  1.5443  ±4.76%     7901   slowest
   · serialize v3 @ 48f1ab4  20,017.24  0.0417  2.0566  0.0500  0.0460  0.1006  0.2656  0.6681  ±1.70%    10009   fastest

 BENCH  Summary

  serialize v0 @ 7529f93 - test/benchmarks.bench.ts > benchmarks > serialize 1 object
    1.05x faster than serialize @ main
    1.05x faster than serialize v3 @ 48f1ab4   
    1.05x faster than serialize v2 @ 73d01b6
    1.06x faster than serialize v1 @ 461e7f3

  serialize @ main - test/benchmarks.bench.ts > benchmarks > serialize array with 1000 identical small objects
    1.00x faster than serialize v0 @ 7529f93
    1.01x faster than serialize v2 @ 73d01b6
    1.02x faster than serialize v3 @ 48f1ab4   
    15.94x faster than serialize v1 @ 461e7f3

  serialize v1 @ 461e7f3 - test/benchmarks.bench.ts > benchmarks > serialize array with 1000 identical small objects (by ref)
    1.05x faster than serialize v0 @ 7529f93
    1.16x faster than serialize v2 @ 73d01b6
    1.20x faster than serialize @ main
    1.23x faster than serialize v3 @ 48f1ab4   

  serialize v0 @ 7529f93 - test/benchmarks.bench.ts > benchmarks > serialize array with 100 identical large circular objects
    1.05x faster than serialize @ main
    1.23x faster than serialize v3 @ 48f1ab4   
    31.38x faster than serialize v2 @ 73d01b6
    36.69x faster than serialize v1 @ 461e7f3

  serialize v0 @ 7529f93 - test/benchmarks.bench.ts > benchmarks > serialize array with 100 identical large objects (by ref)
    1.01x faster than serialize v3 @ 48f1ab4   
    1.05x faster than serialize v2 @ 73d01b6
    1.09x faster than serialize v1 @ 461e7f3
    1.30x faster than serialize @ main

  serialize @ main - test/benchmarks.bench.ts > benchmarks > serialize array with 100 identical large objects
    1.04x faster than serialize v2 @ 73d01b6
    1.06x faster than serialize v0 @ 7529f93
    1.10x faster than serialize v3 @ 48f1ab4   
    20.98x faster than serialize v1 @ 461e7f3

  serialize v3 @ 48f1ab4 - test/benchmarks.bench.ts > benchmarks > serialize array with 100 identical large circular objects (by ref)
    1.05x faster than serialize v0 @ 7529f93
    1.09x faster than serialize @ main
    1.20x faster than serialize v1 @ 461e7f3
    1.27x faster than serialize v2 @ 73d01b6
  ```
</details>

zsilbi commented:
I removed `write()` entirely, it's now as fast (or faster) than `main` in these tests.

```scala
 ✓ test/benchmarks.bench.ts > benchmarks > combined 1282ms
     name                        hz     min      max    mean     p75      p99     p995     p999     rme  samples
   · serialize @ main        119.60  7.3850  10.9026  8.3614  8.6467  10.9026  10.9026  10.9026  ±2.02%       60
   · serialize v3 @ 48f1ab4  132.33  6.7815   9.3977  7.5570  7.8323   9.3977   9.3977   9.3977  ±1.53%       67   fastest

 BENCH  Summary

  serialize v3 @ 48f1ab4 - test/benchmarks.bench.ts > benchmarks > combined
    1.11x faster than serialize @ main
 ```

<details>
  <summary>Detailed results</summary>
 
  ```scala
✓ test/benchmarks.bench.ts > benchmarks > serialize 1 object 1460ms
     name                      hz     min     max    mean     p75     p99    p995    p999     rme  samples
   · serialize @ main  611,269.03  0.0014  0.8672  0.0016  0.0016  0.0027  0.0030  0.0120  ±0.60%   305635
   · serialize v3      623,195.45  0.0014  0.9130  0.0016  0.0016  0.0025  0.0028  0.0084  ±0.56%   311598   fastest

 ✓ test/benchmarks.bench.ts > benchmarks > serialize array with 1000 identical small objects 1214ms
     name                  hz     min     max    mean     p75     p99    p995    p999     rme  samples
   · serialize @ main  649.93  1.4175  2.8447  1.5386  1.4885  2.2584  2.3772  2.8447  ±1.46%      325   fastest
   · serialize v3      644.44  1.4138  2.5248  1.5517  1.5035  2.2311  2.3407  2.5248  ±1.44%      323

 ✓ test/benchmarks.bench.ts > benchmarks > serialize array with 1000 identical small objects (by ref) 1211ms
     name                     hz     min     max    mean     p75     p99    p995    p999     rme  samples
   · serialize @ main  14,487.11  0.0635  0.4052  0.0690  0.0670  0.2012  0.2332  0.2697  ±0.63%     7244
   · serialize v3      18,259.90  0.0504  0.4075  0.0548  0.0529  0.0924  0.2089  0.2730  ±0.62%     9130   fastest

 ✓ test/benchmarks.bench.ts > benchmarks > serialize array with 100 identical large circular objects 1248ms
     name                  hz     min     max    mean     p75     p99    p995    p999     rme  samples
   · serialize @ main  226.80  3.6493  6.9682  4.4092  4.8314  6.8586  6.9682  6.9682  ±3.11%      114
   · serialize v3      250.02  3.4680  5.3158  3.9997  4.3827  5.0560  5.3158  5.3158  ±2.03%      126   fastest

 ✓ test/benchmarks.bench.ts > benchmarks > serialize array with 100 identical large objects (by ref) 1217ms
     name                     hz     min     max    mean     p75     p99    p995    p999     rme  samples
   · serialize @ main  25,104.87  0.0330  1.3367  0.0398  0.0369  0.0802  0.1081  0.8289  ±1.92%    12553
   · serialize v3      25,749.48  0.0306  3.5890  0.0388  0.0351  0.0815  0.1030  0.7805  ±2.31%    12875   fastest

 ✓ test/benchmarks.bench.ts > benchmarks > serialize array with 100 identical large objects 1226ms
     name                  hz     min     max    mean     p75     p99    p995    p999     rme  samples
   · serialize @ main  319.27  2.6399  8.4607  3.1321  3.3291  4.4813  8.4607  8.4607  ±2.99%      160   fastest
   · serialize v3      318.83  2.5747  6.4431  3.1365  3.6262  4.8349  6.4431  6.4431  ±3.08%      160

 ✓ test/benchmarks.bench.ts > benchmarks > serialize array with 100 identical large circular objects (by ref) 1214ms
     name                     hz     min     max    mean     p75     p99    p995    p999     rme  samples
   · serialize @ main  19,754.41  0.0419  1.0404  0.0506  0.0471  0.1030  0.1475  0.7836  ±1.75%     9878
   · serialize v3      21,601.89  0.0386  2.3851  0.0463  0.0429  0.0928  0.1256  0.7483  ±1.92%    10801   fastest

 BENCH  Summary

  serialize v3 - test/benchmarks.bench.ts > benchmarks > serialize 1 object
    1.02x faster than serialize @ main

  serialize @ main - test/benchmarks.bench.ts > benchmarks > serialize array with 1000 identical small objects
    1.01x faster than serialize v3

  serialize v3 - test/benchmarks.bench.ts > benchmarks > serialize array with 1000 identical small objects (by ref)
    1.26x faster than serialize @ main

  serialize v3 - test/benchmarks.bench.ts > benchmarks > serialize array with 100 identical large circular objects
    1.10x faster than serialize @ main

  serialize v3 - test/benchmarks.bench.ts > benchmarks > serialize array with 100 identical large objects (by ref)
    1.03x faster than serialize @ main

  serialize @ main - test/benchmarks.bench.ts > benchmarks > serialize array with 100 identical large objects
    1.00x faster than serialize v3

  serialize v3 - test/benchmarks.bench.ts > benchmarks > serialize array with 100 identical large circular objects (by ref)
    1.09x faster than serialize @ main
  ```
</details>

pi0 commented on src/serialize.ts:
```diff
@@ -101,78 +68,113 @@ const Serializer = /*@__PURE__*/ (function () {
+    $object(object: any) {
+      let content = this.#context.get(object);
+
+      if (content === undefined) {
```
code style: we could short-circuit return here with !== undefined

pi0 replied:
but also extra bytes for double return..

pi0 reviewed (approved):
Love the last refactors with explicit return is much cleaner (no wonder why faster!)
